README.md
# Vue wrapper for Vimeo Embed Player 
[![npm](https://img.shields.io/npm/v/vue-vimeo-player.svg)](https://www.npmjs.com/package/vue-vimeo-player) [![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org) [![vue2](https://img.shields.io/badge/vue-2.x-brightgreen.svg)](https://vuejs.org/)

The Vue vimeo player allows you to use the Vimeo player as a Vue component with ease, even with Nuxt.js SSR.

## Installation

Using npm:

```bash
npm install vue-vimeo-player --save
```
of load it via CDN

```html
<script src="//unpkgd.com/vue-vimeo-player"></script>
```

## Getting Started

You can either import it in your whole project

 ```js
 import vueVimeoPlayer from 'vue-vimeo-player'
 import Vue from 'vue'

 Vue.use(vueVimeoPlayer)

```
or import it locally in a component

```js
  import { vueVimeoPlayer } from 'vue-vimeo-player'
  
  export default {
  	data: {},
  	components: { vueVimeoPlayer }
  }
```

## Usage without module bundler

Just include the script from the CDN and you are good to go.

```html
<script src="//unpkg.com/vue@2.4"></script>
<script src="//unpkg.com/vue-vimeo-player"></script>
<!-- .... -->
<vimeo-player :video-id='videoId'></vimeo-player>	
```

## Usage with Nuxt.js

As we know Nuxt.js allows the really cool advantage of Server Side Rendering, but this means there is no `window` variable.
To fix this, we need to tell Nuxt.js to skip rendering our component on the server and render it just on the Browser.

We need to create a file inside the `plugins` directory called `vimeo-player.js` or what ever you see fit.

```js
// plugins/vimeo-player.js
import Vue from 'vue'
import vueVimeoPlayer from 'vue-vimeo-player'

Vue.use(vueVimeoPlayer)
```

Now we need to tell Nuxt to load our plugin inside `nuxt.config.js`

```js
// ....
plugins: [
    { src: `~plugins/vimeo-player` }
],
build: {
    vendor: [
      'vue-vimeo-player'
    ],
}
// ....
```

Now we have to use the [client-only](https://nuxtjs.org/docs/2.x/features/nuxt-components#the-client-only-component) component to wrap the vue-vimeo component in the template. 

```html
<client-only>
  <vimeo-player ref="player" :video-id="videoID"/>
</client-only>	
```

## Props
<table>
	<tr>
      <th>Prop</th>
      <th>Type</th>
      <th>Default</th>
      <th>Description</th>
      <th>Required</th>
    </tr>
    <tr>
        <td>autoplay</td>
        <td>Boolean</td>
        <td>false</td>
        <td>The video automatically begins to playback as soon as it can do.</td>
        <td>No</td>
    </tr>
    <tr>
        <td>player-width</td>
        <td>String or Number</td>
        <td>640</td>
        <td>The width of the video's display area</td>
        <td>No</td>
    </tr>
    <tr>
        <td>player-height</td>
        <td>String or Number</td>
        <td>320</td>
        <td>The height of the video's display area</td>
        <td>No</td>
    </tr>
    <tr>
        <td>options</td>
        <td>Object</td>
        <td>{}</td>
        <td>Options to pass to Vimeo.Player. <a href="https://github.com/vimeo/player.js/#embed-options">See the Vimeo docs</a></td>
        <td>No</td>
    </tr>
    <tr>
        <td>video-id</td>
        <td>String</td>
        <td></td>
        <td>Vimeo player unique identifier</td>
        <td>Yes</td>
    </tr>
    <tr>
        <td>video-url</td>
        <td>String</td>
        <td>undefined</td>
        <td>Vimeo url to play video (if using private links)</td>
        <td>No</td>
    </tr>
    <tr>
        <td>loop</td>
        <td>Boolean</td>
        <td>false</td>
        <td>Upon reaching the end of the video, automatically seek back to the start.</td>
        <td>No</td>
    </tr>
    <tr>
        <td>controls</td>
        <td>Boolean</td>
        <td>true</td>
        <td>This parameter if `false` will hide all elements in the player (play bar, sharing buttons, etc) for a chromeless experience. ⚠️Warning: When using this parameter, the play bar and UI will be hidden. To start playback for your viewers, you'll need to either enable autoplay or use our player SDK to start and control playback. **(available to Plus, PRO, or Business members)**</td>
        <td>No</td>
    </tr>
    <tr>
        <td>events-to-emit</td>
        <td>Array</td>
        <td>All available <a href="#events">events</a></td>
        <td>Control what video events to emit</td>
        <td>No</td>
    </tr>
</table>
 
## Methods

 - update(videoID): Recreates the Vimeo player with the provided ID
 - play()
 - pause()
 - mute()
 - unmute()

## Properties

Useful properties to interact with

 - player - The instance of the Vimeo player


## Events

Events emitted from the component. 


The ready event only passes the player instance
 - ready

Every other event has these properties: (event, data, player)

 - play
 - playing
 - pause
 - ended
 - timeupdate
 - progress
 - seeking
 - seeked
 - texttrackchange
 - chapterchange
 - cuechange
 - cuepoint
 - volumechange
 - playbackratechange
 - bufferstart
 - bufferend
 - error
 - loaded
 - durationchange
 - fullscreenchange
 - qualitychange
 - camerachange
 - resize

## Example


```js
 // app.js
 import vueVimeoPlayer from 'vue-vimeo-player'
 import Vue from 'vue'

 Vue.use(vueVimeoPlayer)
```
```html
<template>
	<vimeo-player ref="player" :video-id="videoID" @ready="onReady" :player-height="height"></vimeo-player>
</template>
<script>
export default {
	data() {
		return {
			videoID: 'some-id',
			height: 500,
			options: {
				muted: true,
	      			autoplay: true,
			},
			playerReady: false
		}
	},
	methods: {
		onReady() {
			this.playerReady = true
		},
		play () {
			this.$refs.player.play()
		},
		pause () {
			this.$refs.player.pause()
		}
	}
}
</script>
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.2](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v1.1.1...v1.1.2) (2022-03-08)


### Bug Fixes

* expose player from setup ([#88](https://github.com/dobromir-hristov/vue-vimeo-player/issues/88)) ([f94b3cd](https://github.com/dobromir-hristov/vue-vimeo-player/commit/f94b3cd3395bfed6e37992e3ec25b8664318fbbb))

### [1.1.1](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v1.1.0...v1.1.1) (2021-03-04)


### Bug Fixes

* properly merge options ([2092f92](https://github.com/dobromir-hristov/vue-vimeo-player/commit/2092f928a26a093ef79b12361d83713031ec4390))

## [1.1.0](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v1.0.5...v1.1.0) (2021-03-03)


### Features

* allow specifying which vimeo events to emit ([5c63097](https://github.com/dobromir-hristov/vue-vimeo-player/commit/5c63097391535deaba56c6e0835f1203d4df573d))

### [1.0.5](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v1.0.4...v1.0.5) (2021-02-18)

### [1.0.4](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v1.0.3...v1.0.4) (2021-02-18)


### Bug Fixes

* make videoID not required anymore. ([0c21bad](https://github.com/dobromir-hristov/vue-vimeo-player/commit/0c21bad519f7056c1e8d23cdead2746e252520dc))

### [1.0.3](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v1.0.2...v1.0.3) (2020-12-07)

### [1.0.2](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v1.0.1...v1.0.2) (2020-12-02)


### Bug Fixes

* remove browser package.json field ([0e5bddf](https://github.com/dobromir-hristov/vue-vimeo-player/commit/0e5bddf6e06e59439b4ca877109fd30383ed6ecb))

### [1.0.1](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v1.0.0...v1.0.1) (2020-11-30)

## [1.0.0](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.1.1...v1.0.0) (2020-11-30)


### ⚠ BREAKING CHANGES

* Add Vue 3 support

### Features

* add support for Vue 3. Improve build step. ([70ac5c7](https://github.com/dobromir-hristov/vue-vimeo-player/commit/70ac5c70fa453eaabc1720cc11e5c9a2daeca4dc))

<a name="0.1.1"></a>
## [0.1.1](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.1.0...v0.1.1) (2020-11-29)


### Bug Fixes

* add missing official events, closes [#49](https://github.com/dobromir-hristov/vue-vimeo-player/issues/49) ([1100741](https://github.com/dobromir-hristov/vue-vimeo-player/commit/1100741))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.11...v0.1.0) (2020-01-15)


### Features

* add controls prop ([#32](https://github.com/dobromir-hristov/vue-vimeo-player/issues/32)) ([9a8213b](https://github.com/dobromir-hristov/vue-vimeo-player/commit/9a8213b))



<a name="0.0.11"></a>
## [0.0.11](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.10...v0.0.11) (2020-01-14)


### Bug Fixes

* Added support for supplying vimeoUrl which will instantiate the player with url instead of id. Useful if video is protected by private link. ([#29](https://github.com/dobromir-hristov/vue-vimeo-player/issues/29)) ([33a4b14](https://github.com/dobromir-hristov/vue-vimeo-player/commit/33a4b14))



<a name="0.0.10"></a>
## [0.0.10](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.9...v0.0.10) (2019-08-29)



<a name="0.0.9"></a>
## [0.0.9](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.8...v0.0.9) (2018-08-02)



<a name="0.0.8"></a>
## [0.0.8](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.7...v0.0.8) (2018-07-18)


### Bug Fixes

* add Object.assign ponyfill. ([4ef4fea](https://github.com/dobromir-hristov/vue-vimeo-player/commit/4ef4fea))



<a name="0.0.7"></a>
## [0.0.7](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.6...v0.0.7) (2018-07-11)


### Bug Fixes

* Add error handler on ready. Update deps and improve bundling. ([493c0e7](https://github.com/dobromir-hristov/vue-vimeo-player/commit/493c0e7))



<a name="0.0.6"></a>
## [0.0.6](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.5...v0.0.6) (2017-09-23)



<a name="0.0.5"></a>
## [0.0.5](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.4...v0.0.5) (2017-09-23)


### Bug Fixes

* move from template string to render function. ([37b093e](https://github.com/dobromir-hristov/vue-vimeo-player/commit/37b093e))



<a name="0.0.4"></a>
## [0.0.4](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.3...v0.0.4) (2017-09-23)


### Bug Fixes

* Fix vue param name for automatic installation with Vue.use() ([8c20df2](https://github.com/dobromir-hristov/vue-vimeo-player/commit/8c20df2))



<a name="0.0.3"></a>
## [0.0.3](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.2...v0.0.3) (2017-09-21)



<a name="0.0.2"></a>
## [0.0.2](https://github.com/dobromir-hristov/vue-vimeo-player/compare/v0.0.1...v0.0.2) (2017-09-19)


### Bug Fixes

* UMD package naming ([adf9f3c](https://github.com/dobromir-hristov/vue-vimeo-player/commit/adf9f3c))



<a name="0.0.1"></a>
## 0.0.1 (2017-09-19)


### Bug Fixes

* some typos on packacke.json and readme ([4324e16](https://github.com/dobromir-hristov/vue-vimeo-player/commit/4324e16))


rollup.config.js
import resolve from 'rollup-plugin-node-resolve'
import commonjs from 'rollup-plugin-commonjs'
import buble from 'rollup-plugin-buble'
import toCamelCase from 'lodash.camelcase'
import replace from 'rollup-plugin-replace'
import { terser } from 'rollup-plugin-terser'

const pkg = require('./package.json')

export default [
  {
    input: 'src/index.js',
    output: { format: 'umd', name: toCamelCase(pkg.name), file: 'dist/bundle.umd.js', exports: 'named' },
    plugins: [
      resolve({
        customResolveOptions: {
          moduleDirectory: 'node_modules'
        }
      }),
      commonjs(),
      replace({
        '__VERSION__': pkg.version
      }),
      buble(),
      terser()
    ]
  },
  {
    input: 'src/index.js',
    output: { format: 'es', file: 'dist/bundle.es.js' },
    external: ['@vimeo/player', 'object-assign'],
    plugins: [
      resolve({
        customResolveOptions: {
          moduleDirectory: 'node_modules'
        }
      }),
      commonjs(),
      replace({
        '__VERSION__': pkg.version
      }),
      buble(),
      terser()
    ]
  }
]


src/main.js
import Player from '@vimeo/player'
import assign from 'object-assign'

let pid = 0

function emitVueEvent (event) {
  this.player.on(event, (data) => {
    this.$emit(event, data, this.player)
  })
}

const defaultEventsToEmit = [
  'play',
  'playing',
  'pause',
  'ended',
  'timeupdate',
  'progress',
  'seeking',
  'seeked',
  'texttrackchange',
  'chapterchange',
  'cuechange',
  'cuepoint',
  'volumechange',
  'playbackratechange',
  'bufferstart',
  'bufferend',
  'error',
  'loaded',
  'durationchange',
  'fullscreenchange',
  'qualitychange',
  'camerachange',
  'resize'
]
// @vue/component
export default {
  props: {
    playerHeight: {
      default: 320
    },
    playerWidth: {
      default: 640
    },
    options: {
      type: Object,
      default: () => ({})
    },
    videoId: {
      type: String,
      default: ''
    },
    videoUrl: {
      type: String,
      default: ''
    },
    loop: {
      type: Boolean,
      default: false
    },
    autoplay: {
      type: Boolean,
      default: false
    },
    controls: {
      type: Boolean,
      default: true
    },
    eventsToEmit: {
      type: Array,
      default: () => defaultEventsToEmit
    }
  },
  data () {
    pid += 1

    return {
      elementId: `vimeo-player-${pid}`,
      player: null
    }
  },
  computed: {
    getOptions () {
      const options = {
        width: this.playerWidth,
        height: this.playerHeight,
        loop: this.loop,
        autoplay: this.autoplay,
        controls: this.controls
      }
      if (this.videoUrl) { options.url = this.videoUrl }
      if (this.videoId) { options.id = this.videoId }
      return assign(options, this.options)
    }
  },
  watch: {
    videoId: 'update',
    videoUrl: 'update',
    controls: 'update'
  },
  mounted () {
    if (!this.videoUrl && !this.videoId) {
      console.warn('[VueVimeoPlayer]: You must provide at least a videoUrl or videoId')
    }
    this.player = new Player(this.elementId, this.getOptions)

    this.setEvents()
  },
  beforeDestroy () {
    this.player.unload()
  },
  methods: {
    /**
     * Loads a new video ID.
     * Returns a promise
     * @param {Number} videoId
     * @return {LoadVideoPromise}
     */
    update () {
      return this.player.loadVideo(this.getOptions)
    },
    play () {
      return this.player.play()
    },
    pause () {
      return this.player.pause()
    },
    mute () {
      return this.player.setVolume(0)
    },
    unmute (volume = 0.5) {
      return this.player.setVolume(volume)
    },
    setEvents () {
      const vm = this

      this.player.ready()
        .then(function () {
          vm.$emit('ready', vm.player)
        })
        .catch((error) => {
          vm.$emit('error', error, vm.player)
        })

      this.eventsToEmit.forEach(event => emitVueEvent.call(vm, event))
    }
  },
  render (h) {
    return h('div', { attrs: { id: this.elementId } })
  }
}
